Add reference to VNET resource group

ozboms opened the pull request:
When creating a new virtual machine, added new parameter to allow creation of VM in a resource group using virtual network from another resource group

##### SUMMARY
Fixes bug #25032 - azure_rm_virtualmachine fails to find VNET in different resource group 
(https://github.com/ansible/ansible/issues/25032)

Created a new parameter to create an option for another resource group specific to the VNET you want to use called: virtual_network_resource_group. 

Need to input the other resource group where your VNET resides to use another VNET not in the resource group of your new virtual machine. 

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
azure_rm_virtualmachine

ozboms committed:
Add reference to VNET resource group

When creating a new virtual machine, added new parameter to allow creation of VM in a resource group using virtual network from another resource group

harijayms commented:
looks good to me, if there is a test for Virtual machines please add a test as well

ozboms committed:
Fixed white space errors

Was getting compialtion errors with code and just went around and cleaned up extra white spaces and such.

ozboms committed:
Re-fixed more white space issues

tstringer commented:
@ozboms add a `version_added` prop to your new param:

```
    virtual_network_resource_group:
        description:
            - When creating a virtual machine, if a specific virtual network from another resource group should be
              used, use thus parameter to specify the resource group to use.
        default: null
        required: false
        version_added: "2.4"
```

This is explained [here in the docs](http://docs.ansible.com/ansible/dev_guide/developing_modules_documenting.html) (thanks to @gundalow and @jctanner for pointing that out!).

ozboms committed:
Fixed Version Number for new parameter

haroldwongms commented:
This looks good to me.

haroldwongms commented:
shipit

tstringer commented:
shipit

jborean93 commented:
bot_status

s-hertel commented on lib/ansible/modules/cloud/azure/azure_rm_virtualmachine.py:
```diff
@@ -1215,11 +1224,18 @@ def create_default_nic(self):
+                try:
+                    self.network_client.virtual_networks.list(self.virtual_network_resource_group, self.virtual_network_name)
+                    virtual_network_name = self.virtual_network_name
+                except Exception as exc:
```
Can CloudError be caught instead of Exception to be more specific?

trstringer replied:
Yes, completely agreed. @ozboms can you use `CloudError` for your error handling here?

ozboms replied:
Yep! Fixed and recommitted!

s-hertel commented on lib/ansible/modules/cloud/azure/azure_rm_virtualmachine.py:
```diff
@@ -1215,11 +1224,18 @@ def create_default_nic(self):
+                try:
+                    self.network_client.virtual_networks.list(self.resource_group, self.virtual_network_name)
+                    virtual_network_name = self.virtual_network_name
+                except Exception as exc:
```
Here too?

trstringer replied:
Yep, agreed. // @ozboms

ozboms committed:
specified exceptions

Use clouderror instead of exceptions for finidng virtual network

nitzmahone commented on lib/ansible/modules/cloud/azure/azure_rm_virtualmachine.py:
```diff
@@ -1,7 +1,6 @@
 #!/usr/bin/python
 #
-# Copyright (c) 2016 Matt Davis, <mdavis@ansible.com>
```
Why have you removed our copyright and replaced it with your own for making a few lines of change?

ozboms replied:
Oh my bad! I'm creating a few new modules and mistakenly changed this one also in this branch. I apologize for the error and will fix that.

nitzmahone replied:
No worries- we assumed it was an oversight. ;)

nitzmahone commented on lib/ansible/modules/cloud/azure/azure_rm_virtualmachine.py:
```diff
@@ -1253,10 +1268,16 @@ def create_default_nic(self):
-                self.fail(no_subnets_msg)
+            if self.virtual_network_resource_group:
+                try:
+                    subnets = self.network_client.subnets.list(self.virtual_network_resource_group, virtual_network_name)
```
No need to duplicate these lines of code- just conditionally assign a local `virtual_network_resource_group` as `self.resource_group` or `self.virtual_network_resource_group` as appropriate and use it below the conditional.

ozboms replied:
I could have sworn I tried that and it wouldn't let me use a local variable but I didn't really understand how the module worked that well then. I'll test out the local variable and make the necessary changes.

ozboms committed:
Simplified code lines

created a local variable to either use the VNET resource group or the VM resource group and then used whichever was available as a parameter

ozboms committed:
fixed syntax issue with newline

...hopefully

trstringer commented:
@nitzmahone just revisiting this, are the necessary changes @ozboms made sufficient?

ozboms commented:
@nitzmahone Just wanted to check in to see if my changes are correct and if it's possible to have it merged into the devel branch?

Thanks!

trstringer commented:
@ozboms let's do this. Join up on the IRC meeting tomorrow at 3pm EST (freenode `#ansible-meeting` channel) and let's bring this up then.

@thaumos can you get this PR on the agenda for tomorrow's meeting?

nitzmahone committed:
removed superfluous default/required from docs

nitzmahone committed:
Merge branch 'devel' into ozi_rm_virtualmachine_bugfix

nitzmahone committed:
fix vm_size default

nitzmahone commented:
Sorry, been a hectic couple of weeks- changes look good. I just took the opportunity to get rid of some superfluous `default`/`required` lines (used to be required but aren't anymore). Thanks!

trstringer commented:
Thanks, @nitzmahone!

And thanks for the contribution, @ozboms! 👍 😄

ozboms commented:
Good to know about the default/required lines! 
Thanks for the suggestions and reviewing and approving @nitzmahone @tstringer
